src/lib.rs

```diff
@@ -1,12 +1,43 @@
 #![forbid(unsafe_code)]
+
 pub mod client;
 mod handle;
 mod receive;
 mod send_command;
 mod send_response;
 pub mod server;
+#[cfg(feature = "stream")]
 pub mod stream;
-pub mod types;
-
 #[cfg(test)]
 mod tests;
+pub mod types;
+
+/// Simple abstraction of IMAP flows that can be used for implementing IO utilities.
+///
+/// The utility [`Stream`](stream::Stream) is using this for implementing the IO bits when
+/// dealing with TCP or TLS.
+pub trait Flow {
+    type Event;
+    type Error;
+
+    fn enqueue_input(&mut self, bytes: &[u8]);
+    fn progress(&mut self) -> Result<Self::Event, FlowInterrupt<Self::Error>>;
+}
+
+/// The IMAP flow was interrupted by an event that needs to be handled externally.
+#[derive(Clone, Debug, Eq, PartialEq)]
```

Comment:
```suggestion
#[must_use = "If the IMAP flow is interrupted the interrupt must be handled. Ignoring this might result in a deadlock on IMAP level"]
#[derive(Clone, Debug, Eq, PartialEq)]
```